File: apps/cyberarche/web/src/lib/api/documents.ts
import { del, get, patch, post } from './http';

export interface Document {
	id: string;
	workspace_id: string;
	title: string;
	parent_id: string | null;
	position: number;
	created_by: string;
	created_at: string;
	updated_at: string;
	trashed: boolean;
	teamspace_id: string | null;
	folder_id?: string | null;
}

export const createDocument = (
	workspaceId: string,
	title = '',
	parentId?: string,
	teamspaceId?: string
) =>
	post<Document>('/api/v1/documents', {
		workspace_id: workspaceId,
		title,
		parent_id: parentId ?? null,
		teamspace_id: teamspaceId ?? null
	});

export const getDocument = (id: string) => get<Document>(`/api/v1/documents/${id}`);

/** One crumb of a document's breadcrumb path. */
export interface PathCrumb {
	kind: 'workspace' | 'teamspace' | 'folder' | 'document';
	id: string;
	label: string;
}

/** The document's breadcrumb path: workspace → teamspace? → folders →
 * ancestor documents → the document itself. */
export const getDocumentPath = (id: string) =>
	get<PathCrumb[]>(`/api/v1/documents/${id}/path`);

/** A document's current block tree (for reading content outside the editor,
 * e.g. exporting a whole teamspace/folder). */
export const documentBlocks = (id: string) =>
	get<{ blocks: import('$lib/editor/registry').BlockData[] }>(
		`/api/v1/documents/${id}/blocks`
	);

/** Title search within a workspace (empty query returns all accessible docs). */
export const searchDocuments = (workspaceId: string, q = '', limit = 50) =>
	get<Document[]>(
		`/api/v1/workspaces/${workspaceId}/search?${new URLSearchParams({ q, limit: String(limit) })}`
	);

/** Documents that reference this one via a [[title]] wikilink. */
export const backlinks = (documentId: string) =>
	get<Document[]>(`/api/v1/documents/${documentId}/backlinks`);

/** Documents in the workspace's trash (soft-deleted, restorable). */
export const listTrashed = (workspaceId: string) =>
	get<Document[]>(`/api/v1/workspaces/${workspaceId}/trash`);

export const listChildren = (workspaceId: string, parentId?: string) => {
	const params = new URLSearchParams({ workspace_id: workspaceId });
	if (parentId) params.set('parent_id', parentId);
	return get<Document[]>(`/api/v1/documents?${params}`);
};

export const retitleDocument = (id: string, title: string) =>
	patch<Document>(`/api/v1/documents/${id}/title`, { title });

export const trashDocument = (id: string) => del<Document>(`/api/v1/documents/${id}`);
export const restoreDocument = (id: string) =>
	post<Document>(`/api/v1/documents/${id}/restore`);
/** Permanently delete a trashed document and its subtree; returns the ids removed. */
export const purgeDocument = (id: string) => del<{ purged: string[] }>(`/api/v1/documents/${id}/trash`);

File: libs/cyberarche/adapters/src/cyberarche/adapters/inbound/http/routers/documents.py
from __future__ import annotations

import base64

from fastapi import APIRouter

from cyberarche.adapters.inbound.http.dependencies import Caller, Cases
from cyberarche.adapters.inbound.http.schemas import (
    CreateDocumentRequest,
    DocumentResponse,
    MoveDocumentRequest,
    PurgeResponse,
    RecordSnapshotRequest,
    RetitleDocumentRequest,
    SnapshotDetailResponse,
    SnapshotResponse,
)
from cyberarche.domain.ids import DocumentId, SnapshotId, TeamspaceId, WorkspaceId

router = APIRouter(prefix="/api/v1/documents", tags=["documents"])


@router.post("", status_code=201)
async def create_document(
    body: CreateDocumentRequest, cases: Cases, caller: Caller
) -> DocumentResponse:
    document = await cases.documents.create(
        caller,
        workspace_id=WorkspaceId(body.workspace_id),
        title=body.title,
        parent_id=DocumentId(body.parent_id) if body.parent_id else None,
        teamspace_id=TeamspaceId(body.teamspace_id) if body.teamspace_id else None,
    )
    return DocumentResponse.from_domain(document)


@router.get("/{document_id}")
async def get_document(document_id: str, cases: Cases, caller: Caller) -> DocumentResponse:
    document = await cases.documents.get(caller, DocumentId(document_id))
    return DocumentResponse.from_domain(document)


@router.get("/{document_id}/backlinks")
async def document_backlinks(
    document_id: str, cases: Cases, caller: Caller
) -> list[DocumentResponse]:
    """Documents that reference this one via a `[[title]]` wikilink."""
    documents = await cases.links.backlinks(caller, DocumentId(document_id))
    return [DocumentResponse.from_domain(d) for d in documents]


@router.get("")
async def list_children(
    workspace_id: str,
    cases: Cases,
    caller: Caller,
    parent_id: str | None = None,
) -> list[DocumentResponse]:
    documents = await cases.documents.children(
        caller,
        workspace_id=WorkspaceId(workspace_id),
        parent_id=DocumentId(parent_id) if parent_id else None,
    )
    return [DocumentResponse.from_domain(d) for d in documents]


@router.patch("/{document_id}/title")
async def retitle_document(
    document_id: str, body: RetitleDocumentRequest, cases: Cases, caller: Caller
) -> DocumentResponse:
    document = await cases.documents.retitle(
        caller, DocumentId(document_id), title=body.title
    )
    return DocumentResponse.from_domain(document)


@router.post("/{document_id}/move")
async def move_document(
    document_id: str, body: MoveDocumentRequest, cases: Cases, caller: Caller
) -> DocumentResponse:
    document = await cases.documents.move(
        caller,
        DocumentId(document_id),
        parent_id=DocumentId(body.parent_id) if body.parent_id else None,
        position=body.position,
    )
    return DocumentResponse.from_domain(document)


@router.delete("/{document_id}")
async def trash_document(document_id: str, cases: Cases, caller: Caller) -> DocumentResponse:
    document = await cases.documents.trash(caller, DocumentId(document_id))
    return DocumentResponse.from_domain(document)


@router.post("/{document_id}/restore")
async def restore_document(
    document_id: str, cases: Cases, caller: Caller
) -> DocumentResponse:
    document = await cases.documents.restore(caller, DocumentId(document_id))
    return DocumentResponse.from_domain(document)


@router.delete("/{document_id}/trash")
async def purge_document(
    document_id: str, cases: Cases, caller: Caller
) -> PurgeResponse:
    """Permanently delete a trashed document and its subtree. The soft-delete
    (move to trash) stays on `DELETE /{document_id}`."""
    purged = await cases.documents.purge(caller, DocumentId(document_id))
    return PurgeResponse(purged=[str(document_id) for document_id in purged])


# ---- Snapshots -------------------------------------------------------------


@router.post("/{document_id}/snapshots", status_code=201)
async def record_snapshot(
    document_id: str, body: RecordSnapshotRequest, cases: Cases, caller: Caller
) -> SnapshotResponse:
    snapshot = await cases.snapshots.record(
        caller,
        DocumentId(document_id),
        content=body.content,
        state_vector=base64.b64decode(body.state_vector_b64 or b""),
    )
    return SnapshotResponse.from_domain(snapshot)


@router.get("/{document_id}/snapshots")
async def list_snapshots(
    document_id: str, cases: Cases, caller: Caller
) -> list[SnapshotResponse]:
    snapshots = await cases.snapshots.list(caller, DocumentId(document_id))
    return [SnapshotResponse.from_domain(s) for s in snapshots]


@router.post("/{document_id}/snapshots/{snapshot_id}/restore")
async def restore_snapshot(
    document_id: str, snapshot_id: str, cases: Cases, caller: Caller
) -> SnapshotDetailResponse:
    snapshot = await cases.snapshots.restore(
        caller, DocumentId(document_id), SnapshotId(snapshot_id)
    )
    return SnapshotDetailResponse.from_domain(snapshot)
